### PR TITLE
[inductor] add decompositions for aten.angle

### LINKED ISSUE #105564: aten.angle
cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by ydwu4**
Fixes #105564.

Added tests.

CPU benchmarking result:
Before decomposition:
```
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]  ===== Forward graph 0 =====
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]  <eval_with_key>.4 from /home/yidi/local/pytorch/torch/fx/experimental/proxy_tensor.py:477 in wrapped class <lambda>(torch.nn.Module):
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, arg0_1: f32[100000]):
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/yidi/local/t.py:5, code: return torch.angle(x)
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]         angle: f32[100000] = torch.ops.aten.angle.default(arg0_1);  arg0_1 = None
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return (angle,)
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO]         
[2023-07-19 14:59:51,277] torch._functorch.aot_autograd.__aot_graphs: [INFO] 
eager:
per-call time (us): 1069.2930221557617
compiled:
per-call time (us): 742.4068450927734
```

After decomposition:
```
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO] TRACED GRAPH
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]  ===== Forward graph 0 =====
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]  <eval_with_key>.4 from /home/yidi/local/pytorch/torch/fx/experimental/proxy_tensor.py:477 in wrapped class <lambda>(torch.nn.Module):
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]     def forward(self, arg0_1: f32[100000]):
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         # File: /home/yidi/local/t.py:5, code: return torch.angle(x)
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         lt: b8[100000] = torch.ops.aten.lt.Scalar(arg0_1, 0)
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         scalar_tensor: f32[] = torch.ops.aten.scalar_tensor.default(0.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         scalar_tensor_1: f32[] = torch.ops.aten.scalar_tensor.default(3.141592653589793, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         where: f32[100000] = torch.ops.aten.where.self(lt, scalar_tensor_1, scalar_tensor);  lt = scalar_tensor_1 = scalar_tensor = None
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         isnan: b8[100000] = torch.ops.aten.isnan.default(arg0_1);  arg0_1 = None
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         scalar_tensor_2: f32[] = torch.ops.aten.scalar_tensor.default(0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         scalar_tensor_3: f32[] = torch.ops.aten.scalar_tensor.default(nan, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         where_1: f32[100000] = torch.ops.aten.where.self(isnan, scalar_tensor_3, scalar_tensor_2);  isnan = scalar_tensor_3 = scalar_tensor_2 = None
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         add: f32[100000] = torch.ops.aten.add.Tensor(where, where_1);  where = where_1 = None
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         return (add,)
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO]         
[2023-07-19 14:57:53,849] torch._functorch.aot_autograd.__aot_graphs: [INFO] 
eager:
per-call time (us): 1228.0082702636719
compiled:
per-call time (us): 83.6038589477539
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -168,6 +168,22 @@ def cat(tensors, dim=0):
...
+    if x.is_complex():
+        return torch.where(
+            torch.isnan(x.real), float("nan"), torch.atan2(x.imag, x.real)
+        )
+    else:
+        # when x is real number
+        #   if x >= 0, return 0
+        #   if x < 0, return pi
+        #   if x is nan, return nan
+        ret = torch.where(x.real < 0, math.pi, 0.0)
```
Comment (jansel) [inline torch/_inductor/decomposition.py:182]:
```suggestion
        ret = torch.where(x < 0, math.pi, 0.0)
```

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
Can you show an example of the output code?  For complex numbers we might just end up using fallbacks.

**Comment by ydwu4** [comment]
> Can you show an example of the output code? For complex numbers we might just end up using fallbacks.

Sure! For below code:
```python
import torch
@torch.compile
def f(x):
    return torch.angle(x)

x = torch.tensor([-1, 1, 0, float("inf"), 1j])
f(x)
```
The output is following. It seems that the it's somehow handled at least for CPU.
```
(pytorch-3.10) [yidi@devgpu018.ftw1 ~/local/pytorch]$ TORCH_LOGS="output_code" python t.py 
/home/yidi/local/pytorch/torch/_inductor/lowering.py:1296: UserWarning: Torchinductor does not support code generation for complex operators. Performance may be worse than eager.
  warnings.warn(
[2023-07-19 16:30:57,985] torch._inductor.graph.__output_code: [INFO] Output code written to: /tmp/torchinductor_yidi/sg/csgfl2ypssdblsuaiktro237jlw3co3hdrzz35uqvdyepvyhfp35.py
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] Output code: 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from ctypes import c_void_p, c_long
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] import torch
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] import math
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] import random
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] import os
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] import tempfile
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from math import inf, nan
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.hooks import run_intermediate_hooks
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.utils import maybe_profile
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from torch import empty_strided, as_strided, device
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.codecache import AsyncCompile
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] from torch._inductor.select_algorithm import extern_kernels
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] aten = torch.ops.aten
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] assert_size_stride = torch._C._dynamo.guards.assert_size_stride
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] async_compile = AsyncCompile()
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] cpp_fused_angle_0 = async_compile.cpp('''
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] #include "/tmp/torchinductor_yidi/zr/czrrhd67iy62iqdam5uwroq4ibq3i5oo4yzl6euetoa7k25vfk35.h"
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] extern "C" void kernel(const float* in_ptr0,
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]                        const float* in_ptr1,
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]                        const float* in_ptr2,
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]                        float* out_ptr0)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] {
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     {
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]         #pragma GCC ivdep
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]         for(long i0=static_cast<long>(0L); i0<static_cast<long>(5L); i0+=static_cast<long>(1L))
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]         {
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp0 = in_ptr0[static_cast<long>(2L*i0)];
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp2 = in_ptr1[static_cast<long>(1L + (2L*i0))];
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp3 = in_ptr2[static_cast<long>(2L*i0)];
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp1 = std::isnan(tmp0);
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp4 = std::atan2(tmp2, tmp3);
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp5 = std::numeric_limits<float>::quiet_NaN();
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             auto tmp6 = tmp1 ? tmp5 : tmp4;
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]             out_ptr0[static_cast<long>(i0)] = tmp6;
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]         }
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     }
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] }
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] ''')
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] async_compile.wait(globals())
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] del async_compile
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] def call(args):
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     arg0_1, = args
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     args.clear()
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(arg0_1, (5, ), (1, ))
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf0 = aten.view_as_real.default(arg0_1)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf1 = buf0
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(buf1, (5, 2), (2, 1))
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     del buf0
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf2 = aten.view_as_real.default(arg0_1)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf3 = buf2
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(buf3, (5, 2), (2, 1))
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     del buf2
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf4 = aten.view_as_real.default(arg0_1)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     del arg0_1
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf5 = buf4
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     assert_size_stride(buf5, (5, 2), (2, 1))
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     del buf4
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     buf6 = empty_strided((5, ), (1, ), device='cpu', dtype=torch.float32)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     cpp_fused_angle_0(c_void_p(buf1.data_ptr()), c_void_p(buf3.data_ptr()), c_void_p(buf5.data_ptr()), c_void_p(buf6.data_ptr()))
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     return (buf6, )
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] def benchmark_compiled_module(times=10, repeat=10):
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     from torch._dynamo.testing import rand_strided
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     from torch._inductor.utils import print_performance
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     arg0_1 = rand_strided((5, ), (1, ), device='cpu', dtype=torch.complex64)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     return print_performance(lambda: call([arg0_1]), times=times, repeat=repeat)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] if __name__ == "__main__":
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     from torch._inductor.utils import compiled_module_main
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG]     compiled_module_main('None', benchmark_compiled_module)
[2023-07-19 16:30:57,986] torch._inductor.graph.__output_code: [DEBUG] 
```

**Comment by jansel** [comment]
Looks good, thanks!

**Pull request closed without merging**